set_intersect.c

```diff
@@ -67,33 +63,34 @@ int main()
 /* Merge the second set into the first set;
  * the second set is unchanged.
  */
-static void 
-set_merge(set_t *setA, set_t *setB)
+static set_t *set_intersect(set_t *setA, set_t *setB)
 {
-  set_iterate(setB, setA, set_merge_helper);
+  set_t *set = set_new();
+  set_intersect_data_t *intersect_data = assertp(malloc(sizeof(set_intersect_data_t)), "intersect_data");
+  
+  intersect_data->setA = setA;
+  intersect_data->setB = setB;
+  intersect_data->destination = set;
+
+  set_iterate(setA, intersect_data, set_intersect_helper);
+  set_iterate(setB, intersect_data, set_intersect_helper);
+  
+  return set;
 }
 
-/* Consider one item for insertion into the other set.
- * If the other set does not contain the item, insert it;
- * otherwise, update the other set's item with sum of item values.
+/* Help the set_intersect function to complete it's opperation
  */
-static void 
-set_merge_helper(void *arg, const char *key, void *item)
+static void set_intersect_helper(void *arg, const char *key, void *item)
 {
-  set_t *setA = arg;
-  int *itemB = item;
-  
-  // find the same key in setA
+  set_intersect_data_t *intersect_data = arg;
+  set_t *setA = intersect_data->setA;
+  set_t *setB = intersect_data->setB;
+  set_t *destination = intersect_data->destination;
+
   int *itemA = set_find(setA, key);
-  if (itemA == NULL) {
-    // not found: insert it
-    set_insert(setA, key, intsave(*itemB));
-    printf("\t%s added\n", key);
-  } else {
-    // add to the existing value
-    *itemA += *itemB;
-    printf("\t%s exists\n", key);
-  }
+  int *itemB = set_find(setB, key);
+  if (itemA != NULL && itemB != NULL)
+    set_insert(destination, key, intsave(*itemB));
```

Comment:
Why specifically choose itemB to save in the destination?

Reply:
Perhaps we take the min of the two?

Reply:
K